Assets/Scripts/GameScene/PlayerController.cs

```diff
@@ -0,0 +1,136 @@
+    {
+        _playerInput.PlayerContorls.Enable();
+    }
+
+    private void OnDisable()
+    {
+        _playerInput.PlayerContorls.Disable();
+    }
+
+    public Board GetGameBoard()
+    {
+        return AssetDatabase.LoadAssetAtPath<Board>("Assets/Data/Board.asset");
+    }
+
+    
+    private void HandleMovement()
+    {
+        
+        Vector3 moveDirection = new Vector3(Input.GetAxis("Horizontal"), 0f, Input.GetAxis("Vertical"));
+        moveDirection = transform.TransformDirection(moveDirection);
+        Vector3 newPosition = transform.position + moveDirection * Time.deltaTime * _playerSpeed;
+    
+      
+        RaycastHit hit;
+        float raycastDistance = 1f; 
+        if (Physics.Raycast(newPosition, Vector3.down, out hit, raycastDistance))
+        {
+            _characterController.Move(moveDirection * Time.deltaTime * _playerSpeed);
+        }
+        
+    }
+
+    private void HandleRotation()
+    {
+        throw new System.NotImplementedException();
+    }
+    
+    private void HandleCollision()
+    {
+        throw new System.NotImplementedException();
+    }
+
+    private void HandlePhysics()
+    {
+        throw new System.NotImplementedException();
+    }
+    
+    private void HandleMovementWithPlayerInput()
+    {
+        // Vector3 moveDirection = new Vector3(_playerMovement);
+        _playerMovement = transform.TransformDirection(_playerMovement);
+        Vector3 newPosition = transform.position + _playerMovement * Time.deltaTime * _playerSpeed;
+    
+      
+        RaycastHit hit;
+        float raycastDistance = 1f; 
+        if (Physics.Raycast(newPosition, Vector3.down, out hit, raycastDistance))
+        {
+            _characterController.Move(_playerMovement * Time.deltaTime * _playerSpeed);
+        }
+    }
+    private void InitPlayerInput()
+    {
+        _playerInput = new PlayerInput();
+        
+        _playerInput.PlayerContorls.Movement.started += OnMovementInput;
+        _playerInput.PlayerContorls.Movement.performed += OnMovementInput;
+        _playerInput.PlayerContorls.Movement.canceled += OnMovementInput;
+    }
+    
+    private void OnMovementInput(InputAction.CallbackContext context)
+    {
+        _movementInput = context.ReadValue<Vector2>();
+
+        _playerMovement.x = _movementInput.x;
+        _playerMovement.y = 0.0f;
+        _playerMovement.z = _movementInput.y;
+        _isMovementPressed = _movementInput.x != 0 || _movementInput.y != 0;
+        
+    }
+    
+
+    private void HandleRotationWithPlayerInput()
+    {
+        Vector3 playerDirection = new Vector3(_movementInput.x, 0.0f, _movementInput.y);
+        
+
+        Quaternion currentRotation = transform.rotation;
+        if (_isMovementPressed)
+        {
+            Quaternion targetRotation = Quaternion.LookRotation(playerDirection);
+            Quaternion.Slerp(currentRotation, targetRotation, rotationFactorPerFrame * Time.deltaTime);
+        }
+    }
```

Comment:
If you can explain what's going on here I'll let you do it, but if not, please first understand what Quaternion.Slerp means before using it in this way

Reply:
Transform is a Quaternion object.

we use the Slerp function to rotate from the current gameObject.transform to the new direction we go to according to the input we got. The time it takes to finish the rotation is: rotationFactorPerFrame * Time.deltaTime

I can figure out another way if you think this one is just complicated and we can do it better.